File: plugins/libimhex/source/api/imhex_api.cpp
#include <hex/api/imhex_api.hpp>

#include <hex/api/event.hpp>
#include <hex/helpers/shared_data.hpp>

#include <unistd.h>

#include <hex/helpers/logger.hpp>

namespace hex {

    void ImHexApi::Common::sayHello() {
        log::warn("Hello!");
    }

    void ImHexApi::Common::closeImHex(bool noQuestions) {
        EventManager::post<RequestCloseImHex>(noQuestions);
    }

    void ImHexApi::Common::restartImHex() {
        EventManager::post<RequestCloseImHex>(false);
        std::atexit([]{
            execve(SharedData::mainArgv[0], SharedData::mainArgv, nullptr);
        });
    }


    void ImHexApi::Bookmarks::add(Region region, const std::string &name, const std::string &comment, u32 color) {
        Entry entry;

        entry.region = region;

        entry.name.reserve(name.length());
        entry.comment.reserve(comment.length());
        std::copy(name.begin(), name.end(), std::back_inserter(entry.name));
        std::copy(comment.begin(), comment.end(), std::back_inserter(entry.comment));
        entry.locked = false;

        entry.color = color;

        EventManager::post<RequestAddBookmark>(entry);
    }

    void ImHexApi::Bookmarks::add(u64 addr, size_t size, const std::string &name, const std::string &comment, u32 color) {
        Bookmarks::add(Region{addr, size}, name, comment, color);
    }

    std::list<ImHexApi::Bookmarks::Entry>& ImHexApi::Bookmarks::getEntries() {
        return SharedData::bookmarkEntries;
    }


    prv::Provider* ImHexApi::Provider::get() {
        if (!ImHexApi::Provider::isValid())
            return nullptr;

        return SharedData::providers[SharedData::currentProvider];
    }

    const std::vector<prv::Provider*>& ImHexApi::Provider::getProviders() {
        return SharedData::providers;
    }

    bool ImHexApi::Provider::isValid() {
        return !SharedData::providers.empty();
    }

    void ImHexApi::Provider::add(prv::Provider *provider) {
        SharedData::providers.push_back(provider);
        SharedData::currentProvider = SharedData::providers.size() - 1;
    }

    void ImHexApi::Provider::remove(prv::Provider *provider) {
        auto &providers = SharedData::providers;

        auto it = std::find(providers.begin(), providers.end(), provider);

        providers.erase(it);

        if (it - providers.begin() == SharedData::currentProvider)
            SharedData::currentProvider = 0;

        delete provider;
    }

}

File: plugins/libimhex-rust/src/imhex_api.rs
pub mod ffi {

    pub mod ImHexApi {

        #[cxx::bridge]
        pub mod Common {

            #[namespace = "hex::ImHexApi::Common"]
            extern "C++" {
                include!("hex/api/imhex_api.hpp");

                pub unsafe fn closeImHex(no_questions: bool);
                pub unsafe fn restartImHex();
            }

        }

        #[cxx::bridge]
        pub mod Bookmarks {

            #[namespace = "hex::ImHexApi::Bookmarks"]
            extern "C++" {
                include!("hex/api/imhex_api.hpp");

                pub unsafe fn add(addr : u64, size : usize, name : &CxxString, comment : &CxxString, color : u32);
            }

        }

    }
}

pub struct Region {
    pub address : u64,
    pub size : usize
}

pub struct Color {
    pub a : u8,
    pub g : u8,
    pub b : u8,
    pub r : u8
}

impl Region {

    pub fn new(address : u64, size : usize) -> Self {
        Region { address, size }
    }

}

impl Color {

    pub fn new(r : u8, g : u8, b : u8, a : u8) -> Self {
        Color { a, g, b, r }
    }

    pub fn rgba(self) -> u32 {
        (self.a as u32) << 24 | (self.b as u32) << 16 | (self.g as u32) << 8 | (self.r as u32) << 0
    }
}

pub mod ImHexApi {

    pub mod Common {

        pub fn closeImHex() {

            unsafe {
                crate::imhex_api::ffi::ImHexApi::Common::closeImHex(false);
            }

        }

        pub fn restartImmHex() {

            unsafe {
                crate::imhex_api::ffi::ImHexApi::Common::restartImHex();
            }

        }

    }

    pub mod Bookmarks {

        pub fn add(region : crate::Region, name : &str, comment : &str, color : Option<crate::Color>) {
            cxx::let_cxx_string!(cpp_name = name);
            cxx::let_cxx_string!(cpp_comment = comment);

            unsafe {
                crate::imhex_api::ffi::ImHexApi::Bookmarks::add(region.address, region.size, &cpp_name, &cpp_comment, color.unwrap_or(crate::Color::new(0, 0, 0, 0)).rgba());
            }
        }

    }

}